File: momentra/momentra/Shell/PersonalEmpty/PersonalRelationshipsActivitySheets.swift
import SwiftUI

/// Figma `1036:7697` Recent Activity + `1036:7727` Edit Activity.
struct PersonalRelationshipsActivityFlow: View {
    let momentId: String?
    @Binding var isPresented: Bool
    var onChanged: () -> Void = {}

    @State private var items: [RelationshipsActivityItem] = []
    @State private var filter = "All"
    @State private var editing: RelationshipsActivityItem?

    private let filters = ["All", "Partner", "Family", "Friends", "Self"]
    private let pink = Color(hex: "#E12A9E")
    private let bg = Color(hex: "#14121B")
    private let text = Color(hex: "#E5E0EE")
    private let muted = Color(hex: "#C9C4D8")

    private var filtered: [RelationshipsActivityItem] {
        filter == "All" ? items : items.filter { $0.filter.caseInsensitiveCompare(filter) == .orderedSame }
    }

    var body: some View {
        NavigationStack {
            VStack(spacing: 0) {
                HStack {
                    Button(action: { isPresented = false }) {
                        HStack(spacing: 8) {
                            Text("‹").font(.system(size: 22, weight: .bold))
                                .frame(width: 32, height: 32)
                                .background(Color.white.opacity(0.08))
                                .clipShape(RoundedRectangle(cornerRadius: 8))
                            Text("Recent Activity").font(.system(size: 20, weight: .bold)).foregroundStyle(.white)
                        }
                    }
                    .buttonStyle(.plain)
                    Spacer()
                    Text("🔔").frame(width: 36, height: 36).background(Color.white.opacity(0.06)).clipShape(Circle())
                }
                .padding(.horizontal, 16)
                .padding(.vertical, 12)

                ScrollView(.horizontal, showsIndicators: false) {
                    HStack(spacing: 8) {
                        ForEach(filters, id: \.self) { label in
                            let active = filter == label
                            Text(label)
                                .font(.system(size: 12, weight: .semibold))
                                .foregroundStyle(active ? pink : text)
                                .padding(.horizontal, 12)
                                .padding(.vertical, 6)
                                .background(Color(hex: "#201E28"))
                                .overlay(Capsule().stroke(active ? pink : Color(hex: "#C9BFFF").opacity(0.4)))
                                .clipShape(Capsule())
                                .onTapGesture { filter = label }
                        }
                    }
                    .padding(.horizontal, 16)
                }

                ScrollView {
                    LazyVStack(spacing: 4) {
                        ForEach(filtered) { item in
                            row(item)
                        }
                    }
                    .padding(16)
                }
            }
            .background(bg)
            .task {
                do {
                    let acts = try await APIClient.shared.listPersonalActivity(momentId: momentId, limit: 30)
                    items = RelationshipsActivityModels.from(api: acts)
                } catch {
                    items = []
                }
            }
            .sheet(item: $editing) { item in
                RelationshipsEditActivitySheet(
                    item: item,
                    onClose: { editing = nil },
                    onSave: { updated in
                        items = items.map { $0.id == updated.id ? updated : $0 }
                        editing = nil
                        onChanged()
                    },
                    onDelete: {
                        items.removeAll { $0.id == item.id }
                        editing = nil
                        onChanged()
                    }
                )
                .presentationDetents([.large])
            }
        }
    }

    private func row(_ item: RelationshipsActivityItem) -> some View {
        HStack(spacing: 8) {
            VStack(alignment: .leading, spacing: 2) {
                HStack(spacing: 8) {
                    Text(item.emoji).frame(width: 32, height: 32).background(pink).clipShape(RoundedRectangle(cornerRadius: 16))
                    Text(item.title).font(.system(size: 13, weight: .semibold)).foregroundStyle(text)
                }
                Text(item.whenLabel).font(.system(size: 11)).foregroundStyle(muted)
            }
            Spacer()
            if item.impact.isEmpty {
                Text("Logged")
                    .font(.system(size: 12, weight: .heavy))
                    .foregroundStyle(bg)
                    .padding(.horizontal, 10)
                    .padding(.vertical, 4)
                    .background(pink)
                    .clipShape(Capsule())
            } else {
                Text(item.impact)
                    .font(.system(size: 12, weight: .heavy))
                    .foregroundStyle(bg)
                    .padding(.horizontal, 10)
                    .padding(.vertical, 4)
                    .background(pink)
                    .clipShape(Capsule())
            }
            Button { editing = item } label: {
                Text("✏️").frame(width: 28, height: 28).background(Color(hex: "#2A2834")).clipShape(RoundedRectangle(cornerRadius: 8))
            }
            Button {
                items.removeAll { $0.id == item.id }
                onChanged()
            } label: {
                Text("🗑️").frame(width: 28, height: 28).background(Color(hex: "#3C1E1E")).clipShape(RoundedRectangle(cornerRadius: 8))
            }
        }
        .padding(.vertical, 8)
    }
}

struct RelationshipsEditActivitySheet: View {
    let item: RelationshipsActivityItem
    var onClose: () -> Void
    var onSave: (RelationshipsActivityItem) -> Void
    var onDelete: () -> Void

    @State private var name: String
    @State private var impact: String
    @State private var relationship: String
    @State private var whenLabel: String
    @State private var notes: String
    @State private var selectedTags: Set<String>

    private let chips = ["Romantic", "Supportive", "Playful", "Vulnerable", "Intentional"]
    private let pink = Color(hex: "#E12A9E")
    private let green = Color(hex: "#10B981")

    init(
        item: RelationshipsActivityItem,
        onClose: @escaping () -> Void,
        onSave: @escaping (RelationshipsActivityItem) -> Void,
        onDelete: @escaping () -> Void
    ) {
        self.item = item
        self.onClose = onClose
        self.onSave = onSave
        self.onDelete = onDelete
        _name = State(initialValue: item.title)
        _impact = State(initialValue: item.impact)
        _relationship = State(initialValue: item.relationship)
        _whenLabel = State(initialValue: item.whenLabel)
        _notes = State(initialValue: item.notes.isEmpty
            ? "Spent the evening cooking dinner together and watching a movie. Really felt present and connected. Need to do this more often."
            : item.notes)
        _selectedTags = State(initialValue: Set(item.tags.isEmpty ? ["Romantic"] : item.tags))
    }

    var body: some View {
        ScrollView {
            VStack(spacing: 14) {
                Capsule().fill(Color(hex: "#3A3842")).frame(width: 36, height: 4).padding(.top, 8)
                HStack {
                    Button(action: onClose) {
                        Text("×").font(.system(size: 22)).foregroundStyle(Color(hex: "#FF7A3D"))
                            .frame(width: 40, height: 40)
                            .background(Color.white.opacity(0.06))
                            .clipShape(Circle())
                    }
                    VStack(spacing: 4) {
                        Text("Edit Activity").font(.system(size: 18, weight: .semibold)).foregroundStyle(.white)
                        Text("Edit Entry Details").font(.system(size: 12)).foregroundStyle(Color(hex: "#94A3B8"))
                    }
                    .frame(maxWidth: .infinity)
                    Color.clear.frame(width: 40, height: 40)
                }

                field("ACTIVITY NAME", text: $name, focused: true)
                HStack(spacing: 12) {
                    field("IMPACT", text: $impact, focused: true)
                    VStack(alignment: .leading, spacing: 6) {
                        Text("RELATIONSHIP").font(.system(size: 11, weight: .semibold)).foregroundStyle(Color(hex: "#64748B"))
                        Picker("", selection: $relationship) {
                            ForEach(["Partner", "Family", "Friends", "Self"], id: \.self) { Text($0).tag($0) }
                        }
                        .pickerStyle(.menu)
                        .padding(.horizontal, 10)
                        .padding(.vertical, 8)
                        .frame(maxWidth: .infinity, alignment: .leading)
                        .background(Color(hex: "#3A3842"))
                        .overlay(RoundedRectangle(cornerRadius: 14).stroke(Color(hex: "#938EA1")))
                        .clipShape(RoundedRectangle(cornerRadius: 14))
                    }
                }
                field("DATE & TIME", text: $whenLabel, focused: false)

                VStack(alignment: .leading, spacing: 8) {
                    Text("NOTES").font(.system(size: 11, weight: .semibold)).foregroundStyle(Color(hex: "#64748B"))
                    TextEditor(text: Binding(
                        get: { notes },
                        set: { notes = String($0.prefix(200)) }
                    ))
                    .frame(minHeight: 72)
                    .scrollContentBackground(.hidden)
                    .foregroundStyle(Color(hex: "#E5E0EE"))
                    HStack {
                        Text("Max 200 characters").font(.system(size: 11)).foregroundStyle(Color(hex: "#64748B"))
                        Spacer()
                        Text("\(min(notes.count, 200))/200").font(.system(size: 11, weight: .semibold)).foregroundStyle(green)
                    }
                }
                .padding(12)
                .background(Color(hex: "#14121B"))
                .overlay(RoundedRectangle(cornerRadius: 14).stroke(Color(hex: "#2A2538")))
                .clipShape(RoundedRectangle(cornerRadius: 14))

                FlowLayout(spacing: 8) {
                    ForEach(chips, id: \.self) { chip in
                        let selected = selectedTags.contains(chip)
                        Text(chip)
                            .font(.system(size: 12, weight: .medium))
                            .foregroundStyle(selected ? pink : Color(hex: "#9CA3AF"))
                            .padding(.horizontal, 12)
                            .padding(.vertical, 6)
                            .background(selected ? pink.opacity(0.15) : Color(hex: "#2A2538"))
                            .overlay(Capsule().stroke(selected ? pink.opacity(0.25) : .clear))
                            .clipShape(Capsule())
                            .onTapGesture {
                                if selected { selectedTags.remove(chip) } else { selectedTags.insert(chip) }
                            }
                    }
                }

                Button {
                    onSave(RelationshipsActivityItem(
                        id: item.id,
                        title: name,
                        whenLabel: whenLabel,
                        impact: impact,
                        emoji: item.emoji,
                        relationship: relationship,
                        notes: notes,
                        tags: Array(selectedTags),
                        filter: relationship
                    ))
                } label: {
                    Text("Save Changes")
                        .font(.system(size: 17, weight: .bold))
                        .foregroundStyle(Color(hex: "#14121B"))
                        .frame(maxWidth: .infinity)
                        .padding(.vertical, 16)
                        .background(green)
                        .clipShape(Capsule())
                }

                Button("Delete Activity", role: .destructive, action: onDelete)
                    .font(.system(size: 14, weight: .semibold))
            }
            .padding(.horizontal, 16)
            .padding(.bottom, 24)
        }
        .background(Color(hex: "#1C1926"))
    }

    private func field(_ label: String, text: Binding<String>, focused: Bool) -> some View {
        VStack(alignment: .leading, spacing: 6) {
            Text(label).font(.system(size: 11, weight: .semibold)).foregroundStyle(Color(hex: "#64748B"))
            TextField("", text: text)
                .padding(.horizontal, 14)
                .padding(.vertical, 12)
                .background(Color(hex: "#3A3842"))
                .overlay(RoundedRectangle(cornerRadius: 14).stroke(focused ? pink : Color(hex: "#938EA1")))
                .shadow(color: focused ? pink.opacity(0.35) : .clear, radius: 8)
                .clipShape(RoundedRectangle(cornerRadius: 14))
                .foregroundStyle(Color(hex: "#E5E0EE"))
        }
    }
}

File: apk/app/src/main/java/com/example/momentra/ui/shell/personal/relationships/create/PersonalRelationshipsActivitySheets.kt
package com.example.momentra.ui.shell.personal.relationships.create

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.horizontalScroll
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.ExperimentalLayoutApi
import androidx.compose.foundation.layout.FlowRow
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxHeight
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.heightIn
import androidx.compose.foundation.layout.navigationBarsPadding
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.text.BasicTextField
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.DropdownMenu
import androidx.compose.material3.DropdownMenuItem
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.ModalBottomSheet
import androidx.compose.material3.Text
import androidx.compose.material3.rememberModalBottomSheetState
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.draw.drawBehind
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.SolidColor
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.example.momentra.data.repository.PersonalSliceRepository
import com.example.momentra.ui.theme.PlusJakartaSans

private val SheetBg = Color(0xFF14121B)
private val SheetElevated = Color(0xFF1C1926)
private val SheetField = Color(0xFF3A3842)
private val SheetText = Color(0xFFE5E0EE)
private val SheetMuted = Color(0xFFC9C4D8)
private val SheetDim = Color(0xFF64748B)
private val SheetPink = Color(0xFFE12A9E)
private val SheetGreen = Color(0xFF10B981)
private val SheetRed = Color(0xFFF87171)
private val SheetOrange = Color(0xFFFF7A3D)
private val SheetBorder = Color(0xFF2A2538)

private val RelFilters = listOf("All", "Partner", "Family", "Friends", "Self")
private val RelMoodChips = listOf("Romantic", "Supportive", "Playful", "Vulnerable", "Intentional")
private val RelDropdown = listOf("Partner", "Family", "Friends", "Self")

/**
 * Hosts Recent Activity (`1036:7697`) + Edit Activity (`1036:7727`) for Relationships Pulse.
 */
@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun PersonalRelationshipsActivityFlow(
    momentId: String?,
    visible: Boolean,
    onDismiss: () -> Unit,
    onChanged: () -> Unit,
    repository: PersonalSliceRepository = remember { PersonalSliceRepository() },
) {
    if (!visible) return

    var items by remember { mutableStateOf<List<RelationshipsActivityItem>>(emptyList()) }
    var filter by remember { mutableStateOf("All") }
    var editing by remember { mutableStateOf<RelationshipsActivityItem?>(null) }
    val recentSheetState = rememberModalBottomSheetState(skipPartiallyExpanded = true)
    val editSheetState = rememberModalBottomSheetState(skipPartiallyExpanded = true)

    LaunchedEffect(momentId, visible) {
        if (!visible) return@LaunchedEffect
        repository.getActivity(momentId = momentId, limit = 30).fold(
            onSuccess = { items = mapActivityDtosToRelationships(it.items) },
            onFailure = { items = emptyList() },
        )
    }

    ModalBottomSheet(
        onDismissRequest = onDismiss,
        sheetState = recentSheetState,
        containerColor = SheetBg,
        dragHandle = null,
    ) {
        RelationshipsRecentActivityBody(
            items = items,
            filter = filter,
            onFilter = { filter = it },
            onClose = onDismiss,
            onEdit = { editing = it },
            onDelete = { id ->
                items = items.filterNot { it.id == id }
                onChanged()
            },
        )
    }

    editing?.let { item ->
        ModalBottomSheet(
            onDismissRequest = { editing = null },
            sheetState = editSheetState,
            containerColor = SheetElevated,
            dragHandle = null,
        ) {
            RelationshipsEditActivityBody(
                item = item,
                onClose = { editing = null },
                onSave = { updated ->
                    items = items.map { if (it.id == updated.id) updated else it }
                    editing = null
                    onChanged()
                },
                onDelete = {
                    items = items.filterNot { it.id == item.id }
                    editing = null
                    onChanged()
                },
            )
        }
    }
}

@Composable
private fun RelationshipsRecentActivityBody(
    items: List<RelationshipsActivityItem>,
    filter: String,
    onFilter: (String) -> Unit,
    onClose: () -> Unit,
    onEdit: (RelationshipsActivityItem) -> Unit,
    onDelete: (String) -> Unit,
) {
    val filtered = if (filter == "All") items else items.filter { it.filter.equals(filter, ignoreCase = true) }
    Column(
        modifier = Modifier
            .fillMaxWidth()
            .fillMaxHeight(0.92f)
            .navigationBarsPadding(),
    ) {
        Row(
            modifier = Modifier
                .fillMaxWidth()
                .padding(horizontal = 16.dp, vertical = 12.dp),
            horizontalArrangement = Arrangement.SpaceBetween,
            verticalAlignment = Alignment.CenterVertically,
        ) {
            Row(
                horizontalArrangement = Arrangement.spacedBy(8.dp),
                verticalAlignment = Alignment.CenterVertically,
                modifier = Modifier.clickable(onClick = onClose),
            ) {
                Box(
                    modifier = Modifier
                        .size(32.dp)
                        .clip(RoundedCornerShape(8.dp))
                        .background(Color.White.copy(alpha = 0.08f)),
                    contentAlignment = Alignment.Center,
                ) {
                    Text("‹", color = SheetText, fontSize = 20.sp, fontWeight = FontWeight.Bold)
                }
                Text(
                    "Recent Activity",
                    color = Color.White,
                    fontSize = 20.sp,
                    fontWeight = FontWeight.Bold,
                    fontFamily = PlusJakartaSans,
                )
            }
            Box(
                modifier = Modifier
                    .size(36.dp)
                    .clip(RoundedCornerShape(18.dp))
                    .background(Color.White.copy(alpha = 0.06f)),
                contentAlignment = Alignment.Center,
            ) {
                Text("🔔", fontSize = 16.sp)
            }
        }

        Row(
            modifier = Modifier
                .fillMaxWidth()
                .horizontalScroll(rememberScrollState())
                .padding(horizontal = 16.dp),
            horizontalArrangement = Arrangement.spacedBy(8.dp),
        ) {
            RelFilters.forEach { label ->
                val active = filter == label
                Box(
                    modifier = Modifier
                        .clip(RoundedCornerShape(999.dp))
                        .background(Color(0xFF201E28))
                        .border(
                            width = 1.dp,
                            color = if (active) SheetPink else Color(0xFFC9BFFF).copy(alpha = 0.4f),
                            shape = RoundedCornerShape(999.dp),
                        )
                        .clickable { onFilter(label) }
                        .padding(horizontal = 12.dp, vertical = 6.dp),
                ) {
                    Text(
                        label,
                        color = if (active) SheetPink else SheetText,
                        fontSize = 12.sp,
                        fontWeight = FontWeight.SemiBold,
                        fontFamily = PlusJakartaSans,
                    )
                }
            }
        }

        Spacer(Modifier.height(12.dp))

        Column(
            modifier = Modifier
                .weight(1f)
                .verticalScroll(rememberScrollState())
                .padding(horizontal = 16.dp),
            verticalArrangement = Arrangement.spacedBy(4.dp),
        ) {
            filtered.forEach { item ->
                RelationshipsActivityManageRow(
                    item = item,
                    onEdit = { onEdit(item) },
                    onDelete = { onDelete(item.id) },
                )
            }
            Spacer(Modifier.height(24.dp))
        }
    }
}

@Composable
private fun RelationshipsActivityManageRow(
    item: RelationshipsActivityItem,
    onEdit: () -> Unit,
    onDelete: () -> Unit,
) {
    Row(
        modifier = Modifier
            .fillMaxWidth()
            .padding(vertical = 8.dp),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(8.dp),
    ) {
        Column(modifier = Modifier.weight(1f), verticalArrangement = Arrangement.spacedBy(2.dp)) {
            Row(
                horizontalArrangement = Arrangement.spacedBy(8.dp),
                verticalAlignment = Alignment.CenterVertically,
            ) {
                Box(
                    modifier = Modifier
                        .size(32.dp)
                        .clip(RoundedCornerShape(16.dp))
                        .background(SheetPink),
                    contentAlignment = Alignment.Center,
                ) {
                    Text(item.emoji, fontSize = 14.sp)
                }
                Text(
                    item.title,
                    color = SheetText,
                    fontSize = 13.sp,
                    fontWeight = FontWeight.SemiBold,
                    fontFamily = PlusJakartaSans,
                )
            }
            Text(item.whenLabel, color = SheetMuted, fontSize = 11.sp, fontFamily = PlusJakartaSans)
        }
        Box(
            modifier = Modifier
                .clip(RoundedCornerShape(999.dp))
                .background(SheetPink)
                .padding(horizontal = 10.dp, vertical = 4.dp),
        ) {
            Text(
                item.impact,
                color = SheetBg,
                fontSize = 12.sp,
                fontWeight = FontWeight.ExtraBold,
                fontFamily = PlusJakartaSans,
            )
        }
        Box(
            modifier = Modifier
                .size(28.dp)
                .clip(RoundedCornerShape(8.dp))
                .background(Color(0xFF2A2834))
                .clickable(onClick = onEdit),
            contentAlignment = Alignment.Center,
        ) {
            Text("✏️", fontSize = 12.sp)
        }
        Box(
            modifier = Modifier
                .size(28.dp)
                .clip(RoundedCornerShape(8.dp))
                .background(Color(0xFF3C1E1E))
                .clickable(onClick = onDelete),
            contentAlignment = Alignment.Center,
        ) {
            Text("🗑️", fontSize = 12.sp)
        }
    }
}

@OptIn(ExperimentalLayoutApi::class)
@Composable
private fun RelationshipsEditActivityBody(
    item: RelationshipsActivityItem,
    onClose: () -> Unit,
    onSave: (RelationshipsActivityItem) -> Unit,
    onDelete: () -> Unit,
) {
    var name by remember(item.id) { mutableStateOf(item.title) }
    var impact by remember(item.id) { mutableStateOf(item.impact) }
    var relationship by remember(item.id) { mutableStateOf(item.relationship) }
    var whenLabel by remember(item.id) { mutableStateOf(item.whenLabel) }
    var notes by remember(item.id) { mutableStateOf(item.notes.ifBlank { "Spent the evening cooking dinner together and watching a movie. Really felt present and connected. Need to do this more often." }) }
    var selectedTags by remember(item.id) { mutableStateOf(item.tags.ifEmpty { listOf("Romantic") }.toSet()) }
    var relMenu by remember { mutableStateOf(false) }

    Column(
        modifier = Modifier
            .fillMaxWidth()
            .navigationBarsPadding()
            .verticalScroll(rememberScrollState())
            .padding(horizontal = 16.dp)
            .padding(bottom = 24.dp),
        verticalArrangement = Arrangement.spacedBy(14.dp),
        horizontalAlignment = Alignment.CenterHorizontally,
    ) {
        Box(
            modifier = Modifier
                .padding(top = 8.dp)
                .size(width = 36.dp, height = 4.dp)
                .clip(RoundedCornerShape(2.dp))
                .background(Color(0xFF3A3842)),
        )
        Row(
            modifier = Modifier.fillMaxWidth(),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            Box(
                modifier = Modifier
                    .size(40.dp)
                    .clip(RoundedCornerShape(20.dp))
                    .background(Color.White.copy(alpha = 0.06f))
                    .clickable(onClick = onClose),
                contentAlignment = Alignment.Center,
            ) {
                Text("×", color = SheetOrange, fontSize = 22.sp)
            }
            Column(modifier = Modifier.weight(1f), horizontalAlignment = Alignment.CenterHorizontally) {
                Text("Edit Activity", color = Color.White, fontSize = 18.sp, fontWeight = FontWeight.SemiBold, fontFamily = PlusJakartaSans)
                Text("Edit Entry Details", color = Color(0xFF94A3B8), fontSize = 12.sp, fontFamily = PlusJakartaSans)
            }
            Spacer(Modifier.size(40.dp))
        }

        PinkField(label = "ACTIVITY NAME", value = name, onValueChange = { name = it }, focused = true)
        Row(modifier = Modifier.fillMaxWidth(), horizontalArrangement = Arrangement.spacedBy(12.dp)) {
            Box(modifier = Modifier.weight(1f)) {
                PinkField(label = "IMPACT", value = impact, onValueChange = { impact = it }, focused = true)
            }
            Column(modifier = Modifier.weight(1f), verticalArrangement = Arrangement.spacedBy(6.dp)) {
                Text("RELATIONSHIP", color = SheetDim, fontSize = 11.sp, fontWeight = FontWeight.SemiBold, fontFamily = PlusJakartaSans)
                Box {
                    Row(
                        modifier = Modifier
                            .fillMaxWidth()
                            .clip(RoundedCornerShape(14.dp))
                            .background(SheetField)
                            .border(1.dp, Color(0xFF938EA1), RoundedCornerShape(14.dp))
                            .clickable { relMenu = true }
                            .padding(horizontal = 14.dp, vertical = 12.dp),
                        horizontalArrangement = Arrangement.SpaceBetween,
                    ) {
                        Text(relationship, color = Color.White, fontSize = 14.sp, fontFamily = PlusJakartaSans)
                        Text("▼", color = Color(0xFF94A3B8), fontSize = 10.sp)
                    }
                    DropdownMenu(expanded = relMenu, onDismissRequest = { relMenu = false }) {
                        RelDropdown.forEach { opt ->
                            DropdownMenuItem(
                                text = { Text(opt) },
                                onClick = {
                                    relationship = opt
                                    relMenu = false
                                },
                            )
                        }
                    }
                }
            }
        }
        PinkField(label = "DATE & TIME", value = whenLabel, onValueChange = { whenLabel = it }, focused = false)

        Column(
            modifier = Modifier
                .fillMaxWidth()
                .clip(RoundedCornerShape(14.dp))
                .background(SheetBg)
                .border(1.dp, SheetBorder, RoundedCornerShape(14.dp))
                .padding(12.dp),
            verticalArrangement = Arrangement.spacedBy(8.dp),
        ) {
            Text("NOTES", color = SheetDim, fontSize = 11.sp, fontWeight = FontWeight.SemiBold, fontFamily = PlusJakartaSans)
            BasicTextField(
                value = notes.take(200),
                onValueChange = { notes = it.take(200) },
                textStyle = TextStyle(color = SheetText, fontSize = 14.sp, fontFamily = PlusJakartaSans),
                cursorBrush = SolidColor(SheetPink),
                modifier = Modifier
                    .fillMaxWidth()
                    .heightIn(min = 72.dp),
            )
            Row(modifier = Modifier.fillMaxWidth(), horizontalArrangement = Arrangement.SpaceBetween) {
                Text("Max 200 characters", color = SheetDim, fontSize = 11.sp, fontFamily = PlusJakartaSans)
                Text(
                    "${notes.length.coerceAtMost(200)}/200",
                    color = SheetGreen,
                    fontSize = 11.sp,
                    fontWeight = FontWeight.SemiBold,
                    fontFamily = PlusJakartaSans,
                )
            }
        }

        FlowRow(
            modifier = Modifier.fillMaxWidth(),
            horizontalArrangement = Arrangement.spacedBy(8.dp),
            verticalArrangement = Arrangement.spacedBy(8.dp),
        ) {
            RelMoodChips.forEach { chip ->
                val selected = chip in selectedTags
                Box(
                    modifier = Modifier
                        .clip(RoundedCornerShape(20.dp))
                        .background(if (selected) SheetPink.copy(alpha = 0.15f) else Color(0xFF2A2538))
                        .border(
                            1.dp,
                            if (selected) SheetPink.copy(alpha = 0.25f) else Color.Transparent,
                            RoundedCornerShape(20.dp),
                        )
                        .clickable {
                            selectedTags = if (selected) selectedTags - chip else selectedTags + chip
                        }
                        .padding(horizontal = 12.dp, vertical = 6.dp),
                ) {
                    Text(
                        chip,
                        color = if (selected) SheetPink else Color(0xFF9CA3AF),
                        fontSize = 12.sp,
                        fontFamily = PlusJakartaSans,
                    )
                }
            }
        }

        Box(
            modifier = Modifier
                .fillMaxWidth()
                .clip(RoundedCornerShape(999.dp))
                .background(SheetGreen)
                .clickable {
                    onSave(
                        item.copy(
                            title = name,
                            impact = impact,
                            relationship = relationship,
                            filter = relationship,
                            whenLabel = whenLabel,
                            notes = notes,
                            tags = selectedTags.toList(),
                        ),
                    )
                }
                .padding(vertical = 16.dp),
            contentAlignment = Alignment.Center,
        ) {
            Text("Save Changes", color = SheetBg, fontSize = 17.sp, fontWeight = FontWeight.Bold)
        }

        Text(
            "Delete Activity",
            color = SheetRed,
            fontSize = 14.sp,
            fontWeight = FontWeight.SemiBold,
            fontFamily = PlusJakartaSans,
            modifier = Modifier
                .clickable(onClick = onDelete)
                .padding(8.dp),
        )
    }
}

@Composable
private fun PinkField(
    label: String,
    value: String,
    onValueChange: (String) -> Unit,
    focused: Boolean,
) {
    Column(
        modifier = Modifier.fillMaxWidth(),
        verticalArrangement = Arrangement.spacedBy(6.dp),
    ) {
        Text(label, color = SheetDim, fontSize = 11.sp, fontWeight = FontWeight.SemiBold, fontFamily = PlusJakartaSans)
        BasicTextField(
            value = value,
            onValueChange = onValueChange,
            singleLine = true,
            textStyle = TextStyle(color = SheetText, fontSize = 14.sp, fontFamily = PlusJakartaSans),
            cursorBrush = SolidColor(SheetPink),
            modifier = Modifier
                .fillMaxWidth()
                .then(
                    if (focused) {
                        Modifier.drawBehind {
                            drawRoundRect(
                                brush = Brush.radialGradient(
                                    colors = listOf(SheetPink.copy(alpha = 0.35f), Color.Transparent),
                                ),
                            )
                        }
                    } else Modifier,
                )
                .clip(RoundedCornerShape(14.dp))
                .background(SheetField)
                .border(
                    1.dp,
                    if (focused) SheetPink else Color(0xFF938EA1),
                    RoundedCornerShape(14.dp),
                )
                .padding(horizontal = 14.dp, vertical = 12.dp),
        )
    }
}
